-   [NBSplice installation](#nbsplice-installation)
-   [Analysis workflow](#analysis-workflow)
    -   [Input Data](#input-data)
        -   [Transcript expression matrix](#transcript-expression-matrix)
        -   [Gene-isoform relationships matrix](#gene-isoform-relationships-matrix)
    -   [IsoDataSet class](#isodataset-class)
        -   [Basics](#basics)
        -   [Creating an object](#creating-an-object)
        -   [Identifying low-expressed isoforms](#identifying-low-expressed-isoforms)
        -   [Testing differential gene splicing](#testing-differential-gene-splicing)
    -   [NBSpliceRes class](#nbspliceres-class)
        -   [Basics](#basics-1)
        -   [Results exploration](#results-exploration)
-   [NBSplice strategy](#nbsplice-strategy)
    -   [Model](#model)
    -   [Implementation](#implementation)
-   [Session info](#session-info)
-   [References](#references)

NBSplice installation
=====================

`NBSplice` is an R package sitting on GitHub. You could downolad it from the repository and then use `R CMD install`. Another easy way to install it is using the `devtools` package. You can install it from CRAN doing:

``` r
install.packages("devtools")
```

Then, you need to load `devtools`:

``` r
library(devtools)
```

Now, you just use the `install_github("author/package")` command:

``` r
install_github("gamerino/NBSplice")
```

Analysis workflow
=================

Here we show the most basic steps for a differential splicing analysis pipeline using the `NBSplice` R package.

Input Data
----------

To start the analysis, an expression matrix at the isoform/transcript level called `isoCounts`, previously generated by upstream processing tools like *Kallisto* (Bray et al. 2016) or *RSEM* (Li and Dewey 2011), is required. You also need a table describing isoform-gene relationships called `geneIso` and a table of experiment information called `designMatrix`. In particular, this table must contain a column storing the experiment factor which will be evaluated. Here, we assume that this column is called `condition`.

### Transcript expression matrix

As input, the `NBSplice` package expects isoform expression counts obtained from RNA-seq in the form of a matrix of numerical values. In this matrix, the count value found in the *i*-th row and the *j*-th column represents the expression counts assigned to isoform *i* in experimental sample *j*. The values in the matrix should be un-normalized. We suggest the use of the effective counts typically reported by the mentioned quantification tools. It is not necessary to round those because `NBSplice` will do it internally. As an example, we have included an `isoCounts` matrix which is a subset of one of the simulated matrix used to demonstrate `NBSplice` utility.

``` r
data(isoCounts, package="NBSplice")
head(isoCounts)
```

    ##                 C1R1 C1R2 C1R3 C1R4 C2R1 C2R2 C2R3 C2R4
    ## ENST00000358244    0    0    0    0    0    0    0    1
    ## ENST00000360761    3   20   46    0    0    0   27    0
    ## ENST00000416767    4    0    0    0    0    0    0    0
    ## ENST00000429973    0    0   68    0    0    0    0    2
    ## ENST00000442258    0  144   53   55  118   94    1   64
    ## ENST00000503607   11    0   21    0    0   20    0    4

``` r
dim(isoCounts)
```

    ## [1] 3122    8

As can be noted, the `isoCounts` matrix represents an experiment with `8` samples were `3122` isoforms were quantified.

### Gene-isoform relationships matrix

The `geneIso` matrix must have two columns called `isoform_id` and `gene_id` and as many rows as the `isoCounts` matrix. The *i*-th row of the `geneIso` matrix specifies from which gene was originated the *i*-th isoform. This information will be used by `NBSplice` methods for model fitting and hypothesis tests. As an example, we provided a `geneIso` matrix associated to the `isoCounts` object previously loaded.

``` r
data(geneIso, package="NBSplice")
head(geneIso)
```

    ##                         gene_id      isoform_id
    ## ENST00000358244 ENSG00000006283 ENST00000358244
    ## ENST00000360761 ENSG00000006283 ENST00000360761
    ## ENST00000416767 ENSG00000006283 ENST00000416767
    ## ENST00000429973 ENSG00000006283 ENST00000429973
    ## ENST00000442258 ENSG00000006283 ENST00000442258
    ## ENST00000503607 ENSG00000006283 ENST00000503607

``` r
dim(geneIso)
```

    ## [1] 3122    2

This matrix contains annotation information, so, you need to build it everytime that you change your reference transcriptome, downloading the respective information from the database site. An easy way to obtain it is extracting this from the quantification files obtained, for example, with RSEM (first two columns). \#\#\# Experiment data and colName

The `designMatrix` table must contain experiment information relevant for the differential splicing analysis. Its number of columns is variable but it should have as rows as columns the `isoCounts` matrix has. Thus, in the `designMatrix` the *i*-th row specifies experimental information associated to the *i*-th sample. It is worth to mention that the order and names of `isoCounts` columns must be the same of the `designMatrix` rows. Currently, `NBSplice` allows two-levels comparison of single-factor experiments (multiple-factor experiment comparisons are in development). In order to do this, it is necessary to specify the argument `colName` which indicates the name of the `designMatrix` column corresponding to that factor. As an example, we provided the `designMatrix` associated to our `isoCounts` matrix.

``` r
data(designMatrix, package="NBSplice")
head(designMatrix)
```

    ##      samples condition
    ## C1R1    C1R1    Normal
    ## C1R2    C1R2    Normal
    ## C1R3    C1R3    Normal
    ## C1R4    C1R4    Normal
    ## C2R1    C2R1     Tumor
    ## C2R2    C2R2     Tumor

``` r
dim(designMatrix)
```

    ## [1] 8 2

In our case, we must define our `colName` argument indicating that our factor of interest is *condition*.

``` r
colName<-"condition"
levels(designMatrix[,colName])
```

    ## [1] "Normal" "Tumor"

As can be noted, our experiment involved two experimental conditions: *Normal* and *Tumor*. Note that this column of the `designMatrix` is a `factor` with two levels where the first level, commonly taken as the reference, is *Normal*.

IsoDataSet class
----------------

### Basics

`NBSplice` incorporates a new class called `IsoDataSet` to store expression counts and experimental data. An object of this class has the next slots: `counts`, `geneCounts`, `colData`, `isoGeneRel`, `design` and `lowExpIndex`. The first slot will contain the `isoCounts` matrix, whereas, the second one store the expression counts at the gene level. To do this, `NBSplice` incorporates a function which takes every gene of `geneIso` and, for each sample, compute its total expression as the sum of the counts of all its isoforms contained in `isoCounts`. The third and fifth slots correspond to `designMatrix` and `geneIso` objects. The slot called `design` save the formula used to model fitting and hypothesis tests, and is internally defined by `NBSplice` based on the `colName` argument. The last argument, `lowExpIndex`, stores the indexes of low-expressed isoforms. The way to define those isoforms is explained in [Identifying low-expressed isoforms](#identifying-low-expressed-isoforms).

### Creating an object

The typical way to build an `IsoDataSet` object is using its constructor which have the same name as the class. In addition to the four arguments previously defined, it is possible to define the `BPPARAM` argument which must be a `BiocParallelParam` instance defining the parallel back-end to be used (Morgan et al., n.d.).

``` r
myIsoDataSet<-IsoDataSet(isoCounts, designMatrix, colName, geneIso)
```

`NBSplice` also provides several methods for easily object and slots inspections.

``` r
show(myIsoDataSet)
```

    ## IsoDataSet 
    ## Isoform Counts: 
    ##                     C1R1     C1R2     C1R3 C1R4 C2R1 C2R2    C2R3     C2R4
    ## ENST00000358244 0.000000  0.00000  0.00000    0    0    0  0.0000 1.572866
    ## ENST00000360761 4.052778 26.30509 61.88435    0    0    0 36.5994 0.000000
    ## ENST00000416767 5.403704  0.00000  0.00000    0    0    0  0.0000 0.000000
    ## 
    ## Gene Counts: 
    ##                      C1R1      C1R2       C1R3      C1R4      C2R1
    ## ENSG00000006283   36.4750  242.0069  322.87488  182.1134  186.6954
    ## ENSG00000006634  412.0324  143.3628   83.40934  466.9224  233.7150
    ## ENSG00000006704 1410.3667 1121.9123 1296.88076 3245.1795 1185.1700
    ##                      C2R2       C2R3      C2R4
    ## ENSG00000006283  239.2775   86.75413  272.1058
    ## ENSG00000006634  161.4061  132.84227  412.0909
    ## ENSG00000006704 1645.2096 1787.94850 1761.6101
    ## 
    ## Isoform-Gene relationship: 
    ##                         gene_id      isoform_id
    ## ENST00000358244 ENSG00000006283 ENST00000358244
    ## ENST00000360761 ENSG00000006283 ENST00000360761
    ## ENST00000416767 ENSG00000006283 ENST00000416767
    ## 
    ## Design Matrix: 
    ##      samples condition
    ## C1R1    C1R1    Normal
    ## C1R2    C1R2    Normal
    ## C1R3    C1R3    Normal
    ## 
    ## Formula: 
    ##  counts ~ condition + iso + condition:iso
    ## <environment: 0x819f840>

``` r
head(isoCounts(myIsoDataSet))
```

    ##                      C1R1      C1R2     C1R3     C1R4     C2R1      C2R2
    ## ENST00000358244  0.000000   0.00000  0.00000  0.00000   0.0000   0.00000
    ## ENST00000360761  4.052778  26.30509 61.88435  0.00000   0.0000   0.00000
    ## ENST00000416767  5.403704   0.00000  0.00000  0.00000   0.0000   0.00000
    ## ENST00000429973  0.000000   0.00000 91.48122  0.00000   0.0000   0.00000
    ## ENST00000442258  0.000000 189.39668 71.30154 75.31007 163.1856 133.08924
    ## ENST00000503607 14.860186   0.00000 28.25155  0.00000   0.0000  28.31686
    ##                      C2R3       C2R4
    ## ENST00000358244  0.000000   1.572866
    ## ENST00000360761 36.599401   0.000000
    ## ENST00000416767  0.000000   0.000000
    ## ENST00000429973  0.000000   3.145732
    ## ENST00000442258  1.355533 100.663435
    ## ENST00000503607  0.000000   6.291465

Note that the `isoCounts` method returns the `counts` slot but it seems to be different to the `IsoCounts` object previously loaded. This is because this method has an additional logical parameter, `CPM` with default value `TRUE`, useful to indicate if expression counts should be returned in *counts per million* (CPM) scale.

### Identifying low-expressed isoforms

Low expression transcripts have been widely recognized as a source of noise in model fitting and parameter estimation of RNA-seq expression counts. Typically, count thresholds are used to decide if a gene or an isoform has low expression. However, `NBSplice` assumes that an isoform has a low expression if their absolute or relative expression is lower than specifics thresholds. This assumption is due to the package model infers differential splicing by means of changes in relative expression of isoforms. Consequently, a transcript which has a very low relative expression in both conditions could exhibit a small change but enough to be erroneously detected as significant. The `buildLowExpIdx` method is responsible for the idenfitication of low-expressed isoforms. It takes an `IsoDataSet` object and two expression thresholds and creates a low-expression index and stores this index in the `lowExpIndex` object slot. The two thresholds are called `ratioThres` and `countsThres`. The first one of those is used to set a minimum of relative expression that each isoform should achieve in all experimental samples. The second one indicates a threshold in CPM for mean expression across conditions. The method also needs the specification of the `colName` argument, in order to compute mean expressions per factor level, and admits parallelization using the `BPPARAM` argument.

``` r
myIsoDataSet<-buildLowExpIdx(myIsoDataSet, colName, ratioThres = 0.01, countThres = 1)
```

### Testing differential gene splicing

Once an `IsoDataSet` was built and its low-expressed isoforms were identified, you could use the `NBSplice` for differential splicing evaluation using the `NBTest` method. It provides a user-friendly interface between model fitting and hypothesis testing and the user by means a single code line. The theory behind the `NBSplice` model is explained in [NBSplice strategy](#nbsplice-strategy). Briefly, `NBTest` iterates along genes obtaining significance p-values at gene and isoform level indicating the occurrence of differential splicing and significant changes in isoform proportion, respectively. Then, isoform and gene p-values are adjusted. Low-expressed isoforms are ignored for model fitting but their mean relative expression per condition are also computed. The method requires the specification of the `colName`, as well as the statistic assumed for hypothesis test p-values distribution. As the `colName` factor could have more than two levels, it is necessary the specification of an additional argument defining the two levels to be contrasted. By default, the `contrast` argument takes the two first levels of `colName`. `NBTest` returns an `NBSpliceRes` object which stores the obtained results.

``` r
myDSResults<-NBTest(myIsoDataSet, colName, test="F")
```

NBSpliceRes class
-----------------

### Basics

`NBSpliceRes` is another new class provided by NBSplice to facilitate the manipulation and exploration of the differential splicing results obtained using the `NBTest` method. An object of this class has three slots called `results`, `lowExpIndex` and `contrast`. The first slot stores differential splicing results for all isoforms, whereas the second slot keeps the indexes of low-expressed isoforms and the third one contains the evaluated contrast. Specific methods for slot accession are also provided by `NBSplice`

``` r
head(lowExpIndex(myDSResults))
```

    ## [1] 1 2 3 4 5 6

``` r
contrast(myDSResults)
```

    ## [1] "Normal" "Tumor"

``` r
head(results(myDSResults))
```

    ##                iso            gene ratio_Normal ratio_Tumor         odd
    ## 13 ENST00000265728 ENSG00000006634   0.09709374  0.15200420  0.44823126
    ## 19 ENST00000265755 ENSG00000006704   0.04243533  0.08382077  0.68069969
    ## 20 ENST00000424337 ENSG00000006704   0.26147010  0.19001001 -0.31924315
    ## 23 ENST00000476977 ENSG00000006704   0.60918127  0.63195494  0.03670222
    ## 36 ENST00000344419 ENSG00000009765   0.64420873  0.58457743 -0.09713355
    ## 40 ENST00000500320 ENSG00000009765   0.29988829  0.29554796 -0.01457894
    ##            stat        pval    genePval        FDR    geneFDR
    ## 13 1.517719e+00 0.264051826 0.264051826 0.73845145 0.63203894
    ## 19 1.240094e+01 0.002437532 0.007312596 0.03718269 0.05101811
    ## 20 3.081323e+00 0.096199912 0.007312596 0.45436349 0.05101811
    ## 23 4.193792e-02 0.840036032 0.007312596 1.00000000 0.05101811
    ## 36 4.425636e-02 0.836907109 0.975506273 1.00000000 1.00000000
    ## 40 9.827851e-04 0.975506273 0.975506273 1.00000000 1.00000000

The `data.frame` containing differential splicing results has the next columns:

-   `iso`: Isoform name.
-   `gene`: Gene name.
-   `ratio_X`: Relative expression of the isoform in experimental condition *X*, which is the first value specified with the `contrast` argument.
-   `ratio_Y`: Relative expression of the isoform in experimental condition *Y*, which is the second value specified with the `contrast` argument.
-   `odd`: Ratio between isoform's relative expression in condition *X* and *Y*.
-   `stat`: Statistic of the hypothesis test performed to decide if the change in isoform's relative expression between *X* and *Y* conditions was significant.
-   `pval`: P-value at the isoform level corresponding to the statistic `stat`.
-   `genePval`: P-value at the gene level.
-   `FDR`: Adjusted p-value at the isoform level.
-   `genePval`: Adjusted p-value at the gene level.

In particular, the `pval` or its adjusted value, `FDR`, is useful for deciding if an isoform experimented significant change in its relative expression due to the change between *X* and *Y* conditions, here called *Normal* and *Tumor*. Whereas, the `genePval` or its adjusted value, `geneFDR`, will help you to decide if a gene evidenced significant changes in its alternative splicing pattern when conditions *X* and *Y* were contrasted.

In the example illustrated above, non-reliable isoforms were discarded by means of the `filter` argument of the `results` methods which has `TRUE` as its default value. Non-reliable refers to both low expressed isoforms and those cases where model estimations did not achieve convergence. To obtain the full `results` matrix only you need specifies that `filter` argument is `FALSE`.

``` r
head(results(myDSResults, filter = FALSE))
```

    ##               iso            gene ratio_Normal ratio_Tumor odd stat pval
    ## 1 ENST00000358244 ENSG00000006283   0.00000000 0.001445087  NA   NA   NA
    ## 2 ENST00000360761 ENSG00000006283   0.10286836 0.105468750  NA   NA   NA
    ## 3 ENST00000416767 ENSG00000006283   0.03703704 0.000000000  NA   NA   NA
    ## 4 ENST00000429973 ENSG00000006283   0.07083333 0.002890173  NA   NA   NA
    ## 5 ENST00000442258 ENSG00000006283   0.35424397 0.453963572  NA   NA   NA
    ## 6 ENST00000503607 ENSG00000006283   0.12372685 0.035366146  NA   NA   NA
    ##   genePval FDR geneFDR
    ## 1       NA  NA      NA
    ## 2       NA  NA      NA
    ## 3       NA  NA      NA
    ## 4       NA  NA      NA
    ## 5       NA  NA      NA
    ## 6       NA  NA      NA

Non-reliable isoforms could be easily detected in the obtained matrix because they present `NA` value in all parameters derived from model fitting and hypothesis test. It is worth to note that although non-reliable isoforms were not analyzed by statistical models, they have been used to compute the total gene counts so in order to avoid overestimation of relative expression of the well-expressed isoforms.

### Results exploration

The `NBSpliceRes` class has several methods useful for differential splicing results exploration.

#### Getting results of differentially spliced genes

`NBSplice` assumes a gene is differentially spliced if its p-value (raw or adjusted) is lower than a significant threshold (default value = 0.05). The results table only for those differentially spliced genes is easily extracted using the `GetDSResults` method. Whereas, the names of those genes is obtained by means of the `GetDSGenes`. Both methods expect at least one argument of class `NBSpliceRes`. In addition, they could receive a logical parameter, `adjusted`, indicating if p-values must be adjusted (default) or not, and a numeric parameter, `p.value`, establishing the significance threshold (default value = 0.05).

``` r
mySignificantRes<-GetDSResults(myDSResults)
head(mySignificantRes)
```

    ##                iso            gene ratio_Normal ratio_Tumor          odd
    ## 19 ENST00000265755 ENSG00000006704   0.04243533  0.08382077  0.680699686
    ## 20 ENST00000424337 ENSG00000006704   0.26147010  0.19001001 -0.319243154
    ## 23 ENST00000476977 ENSG00000006704   0.60918127  0.63195494  0.036702216
    ## 81 ENST00000412343 ENSG00000010256   0.07604155  0.03432414 -0.795430961
    ## 82 ENST00000415995 ENSG00000010256   0.33479692  0.53297672  0.464953609
    ## 83 ENST00000460105 ENSG00000010256   0.12564045  0.12507306 -0.004526187
    ##            stat         pval     genePval        FDR     geneFDR
    ## 19 1.240094e+01 0.0024375319 0.0073125956 0.03718269 0.051018109
    ## 20 3.081323e+00 0.0961999117 0.0073125956 0.45436349 0.051018109
    ## 23 4.193792e-02 0.8400360323 0.0073125956 1.00000000 0.051018109
    ## 81 1.440928e+01 0.0003741743 0.0007471205 0.01161230 0.009274137
    ## 82 5.112639e+00 0.0278036706 0.0007471205 0.19859765 0.009274137
    ## 83 4.799817e-04 0.9826017122 0.0007471205 1.00000000 0.009274137

``` r
dim(mySignificantRes)
```

    ## [1] 184  10

``` r
myDSGenes<-GetDSGenes(myDSResults)
head(myDSGenes)
```

    ## [1] "ENSG00000010256" "ENSG00000039068" "ENSG00000065809" "ENSG00000079999"
    ## [5] "ENSG00000115648" "ENSG00000122085"

``` r
length(myDSGenes)
```

    ## [1] 40

As can be seen, in our example, `NBSplice` detected 184 isoforms with a significant change of relative expression between *Normal* and *Tumor* condition whereas 40 genes were identified as differentially spliced.

#### Isoforms' relative expression plot

The scatter plot of isoforms' relative expression between the two contrasted conditions is explored using the `plotRatiosDisp`. It returns a `ggplot2` object which could be then easily modified and customized.

``` r
plotRatiosDisp(myDSResults)
```

![](NBSplice-vignette_files/figure-markdown_github/plotD-1.png)

As can be noted, isoforms are colored according to if they came from those genes detected as differentially spliced or not. To do this, the `adjusted` and `p.value` parameters, previously used to call the `getDSResults` method, could be specified in the `plotRatiosDisp` calling.

#### Volcano plot

The volcano plot is commonly used to explore the dispersion between statistical significance from a statistical test with the magnitude of the change. In the case of `NBSplice` the isoform p-value and the difference between isoform's relative expression per contrasted condition are explored using the `plotVolcano` method. In addition, isoforms are colored according to if they came from those genes detected as differentially spliced or not.

``` r
plotVolcano(myDSResults)
```

![](NBSplice-vignette_files/figure-markdown_github/plotv-1.png)

#### Exploring a differentially spliced gene

`NBSplice` also provides two methods to individually explore differentially spliced genes: `GetGeneResults` and `plotGeneResults`. The first one is useful for obtaining the differential expression results of a single gene, specified using the `gene` argument. The second one is a graphical method to generate a bar-plot illustrating the isoforms' relative expression values in the two contrasted conditions. In both cases, it is possible filtering those non-reliable isoforms by means of the `filterLowExpIso` argument. In the case of the `plotGeneResults` method, non-significant isoforms could be also filtered out setting the `filterNotSignificant` argument to `TRUE` which is its default value. As an example, the results obtained for the first analyzed gene are shown below.

``` r
## Select the first differentially spliced gene
gene<-GetDSGenes(myDSResults)[1]
GetGeneResults(myDSResults, gene)
```

    ##                iso            gene ratio_Normal ratio_Tumor          odd
    ## 81 ENST00000412343 ENSG00000010256   0.07604155  0.03432414 -0.795430961
    ## 82 ENST00000415995 ENSG00000010256   0.33479692  0.53297672  0.464953609
    ## 83 ENST00000460105 ENSG00000010256   0.12564045  0.12507306 -0.004526187
    ## 84 ENST00000463708 ENSG00000010256   0.05797983  0.03541777 -0.492881415
    ## 85 ENST00000467690 ENSG00000010256   0.05788541  0.02353249 -0.900083462
    ## 86 ENST00000471189 ENSG00000010256   0.06966719  0.03321337 -0.740777003
    ## 87 ENST00000472438 ENSG00000010256   0.10672054  0.10653902 -0.001702329
    ## 88 ENST00000480561 ENSG00000010256   0.07189439  0.05633448 -0.243891383
    ## 89 ENST00000493806 ENSG00000010256   0.06335483  0.03843817 -0.499700288
    ##            stat         pval     genePval         FDR     geneFDR
    ## 81 1.440928e+01 3.741743e-04 0.0007471205 0.011612304 0.009274137
    ## 82 5.112639e+00 2.780367e-02 0.0007471205 0.198597647 0.009274137
    ## 83 4.799817e-04 9.826017e-01 0.0007471205 1.000000000 0.009274137
    ## 84 5.532489e+00 2.234249e-02 0.0007471205 0.177686556 0.009274137
    ## 85 1.812559e+01 8.301339e-05 0.0007471205 0.004394826 0.009274137
    ## 86 1.247777e+01 8.521519e-04 0.0007471205 0.018524180 0.009274137
    ## 87 6.774745e-05 9.934631e-01 0.0007471205 1.000000000 0.009274137
    ## 88 1.374007e+00 2.462702e-01 0.0007471205 0.714978116 0.009274137
    ## 89 5.704827e+00 2.044378e-02 0.0007471205 0.171957024 0.009274137

``` r
plotGeneResults(myDSResults, gene)
```

![](NBSplice-vignette_files/figure-markdown_github/plotGene-1.png)

``` r
## Keeping non-reliable and non-significant isoforms
plotGeneResults(myDSResults, gene, filterLowExpIso = FALSE, filterNotSignificant = FALSE)
```

![](NBSplice-vignette_files/figure-markdown_github/plotGene2-1.png)

NBSplice strategy
=================

`NBSplice`, is a package developed for differential splicing detection starting from isoform quantification data from RNA-seq experiments. Our tool is based on generalized linear models (GLMs) and uses them to infer significant differences in isoforms relative expression values between two experimental conditions. In addition, `NBSplice` not only reveals what genes were under DS but also reveals the identity of the gene isoforms affected by it as well as their relative expression in each experimental condition.

Model
-----

`NBSplice` is based on the use of GLMs to model isoform expression counts. This approach has been widely used for several differential expression tools like `edgeR` (Robinson, McCarthy, and Smyth 2010) and `DESeq2` (Love, Huber, and Anders 2014) for differential gene expression and `DEXSeq` (Anders, Reyes, and Huber 2012) for differential exon usage analysis. These tools are based in negative binomial (NB) models and, in general, they assume that the counts of the i-th gene/exon in the k-th sequenced sample, denoted as \(y_{ik}\), are a random variable with a NB distribution.
\begin{equation}
y_{ik}\sim NB(\mu_{ik}=p_{ik}s_k, \phi_{i}) 
\end{equation}
This NB distribution is characterized by two parameters: the *mean*, \(\mu_{ik}\), and the *dispersion*, \(\phi_{i}\). Specifically, previously mentioned R packages assume that \(\mu_{ik}\) can be expressed as the product of the real proportion of mRNA fragments of the i-th gene in the k-th sample, \(p_{ik}\), with a factor, \(s_k\), related to the sequencing library size. The parameter \(\phi_{i}\) is directly related to the variance of \(y_{ik}\)
\begin{equation}
V(y_{ik})=\mu_{ik}+\phi_i\mu_{ik}
\label{EQ:var}
\end{equation}
Commonly used methods fit GLMs with logarithmic link and with the elements \(x_{kr}\) of the design matrix to infer differential expression
\begin{equation}
log(p_{ik})=\sum\limits_rx_{kr}\beta_{ir}
\label{EQ:mod}
\end{equation}
In the most simple case, a treatment-control experiment, \(r\) be equal to \(1\) and the design matrix only will have one column where each cell indicates if sample \(k\) was or not treated.

`NBSplice` takes the advantage of the previously described strategy although it proposes an alternative approach. It assumes that isoform expression counts, in CPMs, of the i-th isoform from the j-th gene in the k-th sample, \(y_{ijk}\), follow a NB distribution as
\begin{equation}
y_{ijk}\sim NB(media=p_{ijk}\mu_{jk}, dispersion=\phi_{j}) 
\label{EQ:eq5}
\end{equation}
where \(p_{ijk}\) represents the proportion of the mRNA fragments from the j-th gene (\(\mu_{jk}\)) belonging to the i-th isoform and the \(\phi_{j}\) is the dispersion of the j-th gene, in the k-th sample. In this context, for each isoform, it is possible to predict its mean relative expression or proportion by means of a GLM fit at the level gene according to
\begin{equation}
log(p_{ijk})=x_{r.}\beta_{ij}
\label{EQ:eq6}
\end{equation}
where \(\beta_{ij}\) represents the change in \(p_{ijk}\) due to the effect described by the r-th column of the design matrix **X** (\(x_{r.}\)). Thus, `NBSplice` proposes, unlike other tools such as `DEXSeq` or `Limma` (Ritchie et al. 2015), the adjustment of one model for each gene. In particular, this model must consider the different gene isoforms as effects. In the most simple case, where only one factor (*cond*) with two levels (\(l=1, ~2\)) is analyzed , the previous equation is summarized at:
\begin{equation}
ln(p_{ijl})=\mu_0+iso_{ij}+cond_l+ iso_{ij}:cond_l
\label{EQ:eq7}
\end{equation}
Once coefficient models are estimated, \(\hat{\beta}\), `NBSplice` evaluates if the changes in isoforms proportions due to experimental conditions are significant by means of a linear hypothesis test (Wald test) based on the F-statistic defined below (Greene 2017).
\begin{equation}
H_0:cond_l+ iso_{ij}:cond_l=0 \\
F=\frac{(H\hat{\beta}-q)'(H \hat{cov}(\hat{\beta})H')^{-1}(H\hat{\beta}-q)}{p}
\end{equation}
Here, \(H\) is the contrast matrix, \(q\) is a single element matrix having the value of the right side of the test equation (\(q=0\)), \(\hat{cov}(\hat{\beta})\) is the covariance matrix of \(\hat{\beta}\) and \(p\) is equal to the number of rows of \(H\) (\(p=1\)). In particular, \(H\) is a full rank matrix, conformable with \(\hat{\beta}\) so, when they are multiplied, the linear combination of the left side of the test equation is defined. Thus, under the null hypothesis, the F statistics will have \(F\) distribution with \(1\) and \(n-K\) degrees of freedom. Specifically, \(n\) is the number of samples and \(K\) is the length of \(\beta\). When only one experimental factor with two levels is analyzed (treatment-control case), \(K\) will be equal to the double of the number of analyzed isoforms of gene \(j\), \(I_j\). The comparison of the F statistics against its distribution under null hypothesis provides a p-value for each isoform of gene \(j\).

Given the fact that DS affect the whole gene, once isoform hypothesis tests were performed, `NBSplice` uses the Simes test (Simes 1986) to combine isoform p-values obtaining a gene p-value. Given a family of null hypothesis, \(H_{01}, ..., H_{0I_j}\), to test changes in relative isoform expression of the \(I_j\) isoforms from gene \(j\), the Simes test evaluates only one global null hypothesis
\begin{equation}
H_0=\bigcap\limits_{p=1}^{I_j}H_{0i}
\end{equation}
To do this, the isoform p-values are sorted \(p-value_{(1)j}< \dots < p-value_{(m)j}< \dots <p-value{(I)j}\) and then the Simes statistic, \(T_j\), is defined as
\begin{equation}
T_j=\min \{ \frac{I_j p-value{(m)j}}{m}\}
\label{EQ:eq10}
\end{equation}
Under the null hypothesis, this statistic has uniform distribution, \(U(0,1)\), representing a p-value. The Simes test rejects \(H_0\) if \(T_j\) is lower or equal to the significance threshold \(\alpha\) (Simes 1986). Finally, isoform and gene p-values are corrected using the FDR strategy (Benjamini and Hochberg 1995).

Implementation
--------------

`NBSplice` uses commonly used methods and R packages for classical GLM fit (McCullagh 1984). In particular, the `glm.nb` function of the `MASS` package (Venables and Ripley 2002) is used to model fitting. It was chosen because this function allows the estimation of the dispersion parameter reporting a logical parameter indicating if the model estimation converged or not. This parameter is used by `NBSplice` because if estimations do not converge, thus inferences over the corresponding genes won't be done. In addition, the `lht` function of the `car` package (Fox and Weisberg 2011) was used for linear hypothesis tests and the `simes.test` method from `mppa` package (Rubin-Delanchy and Heard 2014), for the Simes test.

Session info
============

``` r
sessionInfo()
```

    ## R version 3.4.1 (2017-06-30)
    ## Platform: x86_64-pc-linux-gnu (64-bit)
    ## Running under: Ubuntu 16.04.1 LTS
    ## 
    ## Matrix products: default
    ## BLAS: /usr/local/lib/R/lib/libRblas.so
    ## LAPACK: /usr/local/lib/R/lib/libRlapack.so
    ## 
    ## locale:
    ##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
    ##  [3] LC_TIME=es_AR.UTF-8        LC_COLLATE=en_US.UTF-8    
    ##  [5] LC_MONETARY=es_AR.UTF-8    LC_MESSAGES=en_US.UTF-8   
    ##  [7] LC_PAPER=es_AR.UTF-8       LC_NAME=C                 
    ##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
    ## [11] LC_MEASUREMENT=es_AR.UTF-8 LC_IDENTIFICATION=C       
    ## 
    ## attached base packages:
    ## [1] stats     graphics  grDevices utils     datasets  methods   base     
    ## 
    ## other attached packages:
    ## [1] NBSplice_1.0.2
    ## 
    ## loaded via a namespace (and not attached):
    ##  [1] Rcpp_0.12.15        mppa_1.0            pillar_1.1.0       
    ##  [4] plyr_1.8.4          compiler_3.4.1      nloptr_1.0.4       
    ##  [7] tools_3.4.1         digest_0.6.15       lme4_1.1-15        
    ## [10] tibble_1.4.2        evaluate_0.10.1     nlme_3.1-131.1     
    ## [13] gtable_0.2.0        lattice_0.20-35     mgcv_1.8-23        
    ## [16] rlang_0.2.0         Matrix_1.2-12       yaml_2.1.16        
    ## [19] parallel_3.4.1      SparseM_1.77        stringr_1.3.0      
    ## [22] knitr_1.20          MatrixModels_0.4-1  locfit_1.5-9.1     
    ## [25] rprojroot_1.3-2     grid_3.4.1          nnet_7.3-12        
    ## [28] BiocParallel_1.10.1 rmarkdown_1.8       limma_3.32.10      
    ## [31] minqa_1.2.4         reshape2_1.4.3      ggplot2_2.2.1      
    ## [34] car_2.1-6           magrittr_1.5        edgeR_3.18.1       
    ## [37] codetools_0.2-15    backports_1.1.2     scales_0.5.0       
    ## [40] htmltools_0.3.6     MASS_7.3-49         splines_3.4.1      
    ## [43] pbkrtest_0.4-7      BiocStyle_2.4.1     colorspace_1.3-2   
    ## [46] labeling_0.3        quantreg_5.35       stringi_1.1.6      
    ## [49] lazyeval_0.2.1      munsell_0.4.3

References
==========

Anders, Simon, Alejandro Reyes, and Wolfgang Huber. 2012. “Detecting Differential Usage of Exons from RNA-seq Data.” *Genome Research* 22 (10). Cold Spring Harbor Lab: 2008–17.

Benjamini, Yoav, and Yosef Hochberg. 1995. “Controlling the False Discovery Rate: A Practical and Powerful Approach to Multiple Testing.” *Journal of the Royal Statistical Society. Series B (Methodological)*. JSTOR, 289–300.

Bray, NL, H Pimentel, P Melsted, and L Pachter. 2016. “Near-Optimal Probabilistic RNA-seq Quantification.” *Nature Biotechnology* 34 (5): 525.

Fox, John, and Sanford Weisberg. 2011. *An R Companion to Applied Regression*. Second. Thousand Oaks CA: Sage. <http://socserv.socsci.mcmaster.ca/jfox/Books/Companion>.

Greene, W.H. 2017. *Econometric Analysis*. Pearson Education. <https://books.google.com.ar/books?id=iICcDgAAQBAJ>.

Li, Bo, and Colin N Dewey. 2011. “RSEM: Accurate Transcript Quantification from RNA-Seq Data with or Without a Reference Genome.” *BMC Bioinformatics* 12 (1). BioMed Central: 323.

Love, Michael I, Wolfgang Huber, and Simon Anders. 2014. “Moderated Estimation of Fold Change and Dispersion for RNA-seq Data with DESeq2.” *Genome Biology* 15 (12). BioMed Central: 550.

McCullagh, Peter. 1984. “Generalized Linear Models.” *European Journal of Operational Research* 16 (3). Elsevier: 285–92.

Morgan, Martin, Valerie Obenchain, Michel Lang, and Ryan Thompson. n.d. *BiocParallel: Bioconductor Facilities for Parallel Evaluation*.

Ritchie, Matthew E, Belinda Phipson, Di Wu, Yifang Hu, Charity W Law, Wei Shi, and Gordon K Smyth. 2015. “Limma Powers Differential Expression Analyses for RNA-Sequencing and Microarray Studies.” *Nucleic Acids Research* 43 (7). Oxford University Press: e47–e47.

Robinson, Mark D, Davis J McCarthy, and Gordon K Smyth. 2010. “edgeR: a Bioconductor Package for Differential Expression Analysis of Digital Gene Expression Data.” *Bioinformatics* 26 (1). Oxford University Press: 139–40.

Rubin-Delanchy, Patrick, and Nicholas A Heard. 2014. *Mppa: Statistics for Analysing Multiple Simultaneous Point Processes on the Real Line*. <https://CRAN.R-project.org/package=mppa>.

Simes, R John. 1986. “An Improved Bonferroni Procedure for Multiple Tests of Significance.” *Biometrika* 73 (3). Oxford University Press: 751–54.

Venables, W. N., and B. D. Ripley. 2002. *Modern Applied Statistics with S*. Fourth. New York: Springer. <http://www.stats.ox.ac.uk/pub/MASS4>.
